prov/shm/src/smr_msg.c

```diff
@@ -192,9 +194,24 @@ static ssize_t smr_generic_sendmsg(struct smr_ep *ep, const struct iovec *iov,
 		}
 		resp = ofi_cirque_tail(smr_resp_queue(ep->region));
 		pend = freestack_pop(ep->pend_fs);
-		smr_format_iov(cmd, smr_peer_addr(ep->region)[peer_id].addr, iov,
-			       iov_count, total_len, op, tag, data, op_flags,
-			       context, ep->region, resp, pend);
+		if (ep->region->cma_cap) {
+			smr_format_iov(cmd, smr_peer_addr(ep->region)[peer_id].addr, iov,
+				       iov_count, total_len, op, tag, data, op_flags,
+				       context, ep->region, resp, pend);
+		} else {
+			msg_id = (ep->msg_id)++;
+			ret = smr_iov_mmap_copy_in(ep, peer_smr, iov, iov_count,
+						   total_len, op, msg_id, &map_name);
+			if (ret) {
+				freestack_push(ep->pend_fs, pend);
+				ret = -FI_EAGAIN;
+				goto unlock_cq;
+			}
+			smr_format_mmap(cmd, smr_peer_addr(ep->region)[peer_id].addr, iov,
+					iov_count, total_len, op, tag, data, op_flags,
+					context, ep->region, msg_id, map_name, resp, pend);
+
+		}
```

Comment:
I'd like to out-dent this and maybe find a way to re-work the if - else if - else protocol selection, so we can prioritize the small transfers and make it clearer that's the purpose of the checks.  As the code is, we have a series of checks and jumps before we can do inline transfers (smr_format_inline), and the code doesn't clearly convey that we're really changing the protocol.

Reply:
I re-ordered to protocol selection, and prioritize inline/inject transfer.